Remove some unneeded CSSValue functions and streamline CSS-wide keyword handling

darinadler opened the pull request:
#### 8e9f9d34d8cfb01b1174c74e39a0fc6effba9343
<pre>
Remove some unneeded CSSValue functions and streamline CSS-wide keyword handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=250513">https://bugs.webkit.org/show_bug.cgi?id=250513</a>
rdar://104167575

Reviewed by Tim Nguyen and Oriol Brufau.

Removed some other unneeded things as as well. One benefit of this change is that the CSSValue
class is a little less cluttered. Another is that the new implementations are mostly inlined
so there might be a small performance boost. Other small improvements in specific code paths.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Made CSSParserIdioms.h a &quot;private&quot; header
since it&apos;s now included indirectly by WebKitLegacy source files. There are other ways to solve
this problem, but I prefer this one since it doesn&apos;t reduce inlining.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValueInternal): Deleted.
* Source/WebCore/css/CSSComputedStyleDeclaration.h: Removed getPropertyCSSValueInternal.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue): Updated the flag name from m_isImplicit
to m_isImplicitInitialValue.
(WebCore::CSSPrimitiveValue::isCSSWideKeyword const): Deleted.
* Source/WebCore/css/CSSPrimitiveValue.h: Removed isInitialValue, isImplicitInitialValue,
isInheritValue, isUnsetValue, isRevertValue, isRevertLayerValue, and isCSSWideKeyword.

* Source/WebCore/css/CSSStyleDeclaration.h: Removed getPropertyCSSValueInternal.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::treatAsInitialValue): Added. Replaces CSSValue::treatAsInitialValue.
This was the only place that function was used.
(WebCore::CSSToStyleMap::mapFillAttachment): Use treatAsInitialValue.
(WebCore::CSSToStyleMap::mapFillClip): Ditto.
(WebCore::CSSToStyleMap::mapFillComposite): Ditto.
(WebCore::CSSToStyleMap::mapFillBlendMode): Ditto.
(WebCore::CSSToStyleMap::mapFillOrigin): Ditto.
(WebCore::CSSToStyleMap::mapFillImage): Ditto.
(WebCore::CSSToStyleMap::mapFillRepeat): Ditto.
(WebCore::CSSToStyleMap::mapFillSize): Ditto.
(WebCore::CSSToStyleMap::mapFillXPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillYPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillMaskMode): Ditto.
(WebCore::CSSToStyleMap::mapAnimationDelay): Ditto.
(WebCore::CSSToStyleMap::mapAnimationDirection): Ditto.
(WebCore::CSSToStyleMap::mapAnimationDuration): Ditto.
(WebCore::CSSToStyleMap::mapAnimationFillMode): Ditto.
(WebCore::CSSToStyleMap::mapAnimationIterationCount): Ditto.
(WebCore::CSSToStyleMap::mapAnimationName): Ditto.
(WebCore::CSSToStyleMap::mapAnimationPlayState): Ditto.
(WebCore::CSSToStyleMap::mapAnimationProperty): Ditto.
(WebCore::CSSToStyleMap::mapAnimationTimingFunction): Ditto.
(WebCore::CSSToStyleMap::mapAnimationCompositeOperation): Ditto.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::treatAsInheritedValue const): Deleted.
(WebCore::CSSValue::treatAsInitialValue const): Deleted.
(WebCore::CSSValue::isInitialValue const): Deleted.
(WebCore::CSSValue::isImplicitInitialValue const): Deleted.
(WebCore::CSSValue::isInheritValue const): Deleted.
(WebCore::CSSValue::isUnsetValue const): Deleted.
(WebCore::CSSValue::isRevertValue const): Deleted.
(WebCore::CSSValue::isRevertLayerValue const): Deleted.
(WebCore::CSSValue::isCSSWideKeyword const): Deleted.

* Source/WebCore/css/CSSValue.h: Use the m_isImplicitInitialValue flag to
implement the isImplicitInitialValue function. Removed isInheritValue,
isInitialValue, isUnsetValue, isRevertValue, isRevertLayerValue,
isCSSWideKeyword, treatAsInitialValue, and treatAsInheritedValue.

* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.h: Use isCSSWideKeyword from CSSParserIdioms.

* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
(WebCore::DeprecatedCSSOMComplexValue::cssValueType const): Use valueID
instead of separate getters.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::getPropertyCSSValue): Call
the underlying getPropertyCSSValue directly instead of through getPropertyCSSValueInternal.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyCSSValueInternal): Deleted.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.h: Removed getPropertyCSSValueInternal.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const): Use isCSSWideKeyword from CSSParserIdioms.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueTextForLonghand): Removed unneeded std::monostate case.

* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::fontWeightValueIsBold): Use isCSSWideKeyword from CSSParserIdioms.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverterCaches::propertyValueForNode): Use isValueID.
(HTMLConverterCaches::floatPropertyValueForNode): Ditto.
(HTMLConverterCaches::colorPropertyValueForNode): Ditto.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::authoredGridTrackSizes): Get the columns or rows property directly from the inline
style rather than going indirectly through cssomStyle, which does the same thing but is less
efficient because it unnecessarily creates an object as a side effect and requires the
getPropertyCSSValueInternal function, not used anywhere else.

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::addMatchedProperties): Use isValueID.
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch): Ditto.

* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty): Use valueID, and simplify since traditional properties
and custom properties are now both expressing special values the same way.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForKeyframe): Use isValueID.
* Source/WebCore/svg/SVGFontFaceElement.cpp:
(WebCore::SVGFontFaceElement::parseAttribute): Ditto.

Canonical link: <a href="https://commits.webkit.org/258863@main">https://commits.webkit.org/258863@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ebaa775b2db382843e1b68bad0da0753abc8658

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12286 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/36172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112410 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172607 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107114 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13307 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3190 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95385 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/110624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10220 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/36172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37848 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/36172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79580 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5695 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/36172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5863 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2812 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11854 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/36172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6092 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7613 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/css/DeprecatedCSSOMValue.cpp:
```diff
@@ -85,16 +85,18 @@ unsigned short DeprecatedCSSOMComplexValue::cssValueType() const
+    case CSSValueInherit:
         return CSS_INHERIT;
-    if (m_value->isInitialValue())
+    case CSSValueInitial:
         return CSS_INITIAL;
-    if (m_value->isUnsetValue())
+    case CSSValueUnset:
         return CSS_UNSET;
-    if (m_value->isRevertValue())
+    case CSSValueRevert:
         return CSS_REVERT;
-    return CSS_CUSTOM;
+    default:
```
Not related to this patch, but this switch isn't handling `revert-layer`, and I don't think the reason is that this is a deprecated class that doesn't need to be updated, since `revert` is present.

weinig replied:
I believe we stopped adding new constants to this class at some point, and that was probably before revert-layer was a thing, so it minding here is probably not something we want to change.

darinadler replied:
Yes, what @weinig said is how I remember it too.

darinadler replied:
Generally speaking we are not expanding the old CSS object model to encompass the many new concepts in modern CSS.

Loirooriol replied:
Sure, just surprising since `revert` is also more modern than the old CSS object model.

Loirooriol commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -593,7 +593,7 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
     if (RefPtr<CSSValue> value = inlineStylePropertyForElement(element, propertyId)) {
         String result;
-        if (value->isInheritValue())
+        if (isValueID(*value, CSSValueInherit))
```
Why not remove `isValueID()` and just use `valueID(*value) == CSSValueInherit`?

darinadler replied:
We could do that, but I thought there was a chance that we might optimize `isValueID` differently at some point, so was leaving it in place for now. It would be a quick mechanical process to remove it, could do it with global replace. I choose not to add to this patch, though.

I think we might want change `CSSValue` to represent value IDs differently. Perhaps just encode the bits right in the `CSSValue` so we can fetch it without checking the type, putting the value `CSSValueInvalid` in for all CSS values that aren’t `CSSPrimitiveValue` containing a value ID. Another possibility would be to encode both the CSS value class and the value ID in a single 16-bit integer; all the value IDs would imply the primitive type. In that case, `isValueID` might be more efficient because it’s comparing with a type that it knows is not `CSSValueInvalid`, whereas `valueID` needs to check and return `CSSValueInvalid`. But also the entire `CSSValue` representation is really left over from when it was the CSS OM object, but that is no longer the case. I wonder why we are reference counting objects and allocating on the heap when so many are just holding an integer.

I also am unclear on why we are using so many different classes that are specific to various types. It seems we need a flexible way to hold the data, a basic object model. The property-specific code can know the shape of the data rather than having to use a property-specific value holder object. We might want to use a scalar-style class like `Color` instead of `Ref<CSSValue>` and do the reference counting internally. It’s great to share the reference counted bit, but not great to allocate on the heap just so we can do polymorphism.

These thoughts are not yet fully formed, but I suspect they could drastically simplify some of our code. My lack of familiarity with the new typed object model is holding me back a little, though.

Not clear what direction we should go, but I do think we might be able to improve efficiency with a better representation.

Loirooriol replied:
That's fair.

darinadler committed:
Remove some unneeded CSSValue functions and streamline CSS-wide keyword handling
https://bugs.webkit.org/show_bug.cgi?id=250513
rdar://104167575

Reviewed by Tim Nguyen and Oriol Brufau.

Removed some other unneeded things as as well. One benefit of this change is that the CSSValue
class is a little less cluttered. Another is that the new implementations are mostly inlined
so there might be a small performance boost. Other small improvements in specific code paths.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Made CSSParserIdioms.h a "private" header
since it's now included indirectly by WebKitLegacy source files. There are other ways to solve
this problem, but I prefer this one since it doesn't reduce inlining.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValueInternal): Deleted.
* Source/WebCore/css/CSSComputedStyleDeclaration.h: Removed getPropertyCSSValueInternal.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue): Updated the flag name from m_isImplicit
to m_isImplicitInitialValue.
(WebCore::CSSPrimitiveValue::isCSSWideKeyword const): Deleted.
* Source/WebCore/css/CSSPrimitiveValue.h: Removed isInitialValue, isImplicitInitialValue,
isInheritValue, isUnsetValue, isRevertValue, isRevertLayerValue, and isCSSWideKeyword.

* Source/WebCore/css/CSSStyleDeclaration.h: Removed getPropertyCSSValueInternal.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::treatAsInitialValue): Added. Replaces CSSValue::treatAsInitialValue.
This was the only place that function was used.
(WebCore::CSSToStyleMap::mapFillAttachment): Use treatAsInitialValue.
(WebCore::CSSToStyleMap::mapFillClip): Ditto.
(WebCore::CSSToStyleMap::mapFillComposite): Ditto.
(WebCore::CSSToStyleMap::mapFillBlendMode): Ditto.
(WebCore::CSSToStyleMap::mapFillOrigin): Ditto.
(WebCore::CSSToStyleMap::mapFillImage): Ditto.
(WebCore::CSSToStyleMap::mapFillRepeat): Ditto.
(WebCore::CSSToStyleMap::mapFillSize): Ditto.
(WebCore::CSSToStyleMap::mapFillXPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillYPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillMaskMode): Ditto.
(WebCore::CSSToStyleMap::mapAnimationDelay): Ditto.
(WebCore::CSSToStyleMap::mapAnimationDirection): Ditto.
(WebCore::CSSToStyleMap::mapAnimationDuration): Ditto.
(WebCore::CSSToStyleMap::mapAnimationFillMode): Ditto.
(WebCore::CSSToStyleMap::mapAnimationIterationCount): Ditto.
(WebCore::CSSToStyleMap::mapAnimationName): Ditto.
(WebCore::CSSToStyleMap::mapAnimationPlayState): Ditto.
(WebCore::CSSToStyleMap::mapAnimationProperty): Ditto.
(WebCore::CSSToStyleMap::mapAnimationTimingFunction): Ditto.
(WebCore::CSSToStyleMap::mapAnimationCompositeOperation): Ditto.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::treatAsInheritedValue const): Deleted.
(WebCore::CSSValue::treatAsInitialValue const): Deleted.
(WebCore::CSSValue::isInitialValue const): Deleted.
(WebCore::CSSValue::isImplicitInitialValue const): Deleted.
(WebCore::CSSValue::isInheritValue const): Deleted.
(WebCore::CSSValue::isUnsetValue const): Deleted.
(WebCore::CSSValue::isRevertValue const): Deleted.
(WebCore::CSSValue::isRevertLayerValue const): Deleted.
(WebCore::CSSValue::isCSSWideKeyword const): Deleted.

* Source/WebCore/css/CSSValue.h: Use the m_isImplicitInitialValue flag to
implement the isImplicitInitialValue function. Removed isInheritValue,
isInitialValue, isUnsetValue, isRevertValue, isRevertLayerValue,
isCSSWideKeyword, treatAsInitialValue, and treatAsInheritedValue.

* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.h: Use isCSSWideKeyword from CSSParserIdioms.

* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
(WebCore::DeprecatedCSSOMComplexValue::cssValueType const): Use valueID
instead of separate getters.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::getPropertyCSSValue): Call
the underlying getPropertyCSSValue directly instead of through getPropertyCSSValueInternal.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyCSSValueInternal): Deleted.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.h: Removed getPropertyCSSValueInternal.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const): Use isCSSWideKeyword from CSSParserIdioms.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueTextForLonghand): Removed unneeded std::monostate case.

* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::fontWeightValueIsBold): Use isCSSWideKeyword from CSSParserIdioms.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverterCaches::propertyValueForNode): Use isValueID.
(HTMLConverterCaches::floatPropertyValueForNode): Ditto.
(HTMLConverterCaches::colorPropertyValueForNode): Ditto.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::authoredGridTrackSizes): Get the columns or rows property directly from the inline
style rather than going indirectly through cssomStyle, which does the same thing but is less
efficient because it unnecessarily creates an object as a side effect and requires the
getPropertyCSSValueInternal function, not used anywhere else.

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::addMatchedProperties): Use isValueID.
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch): Ditto.

* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty): Use valueID, and simplify since traditional properties
and custom properties are now both expressing special values the same way.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForKeyframe): Use isValueID.
* Source/WebCore/svg/SVGFontFaceElement.cpp:
(WebCore::SVGFontFaceElement::parseAttribute): Ditto.

Canonical link: https://commits.webkit.org/258863@main
